ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java

```diff
@@ -84,23 +84,13 @@ public static UnsignedLong compute_slots_since_epoch_start(UnsignedLong slot) {
    */
   private static Optional<Bytes32> get_ancestor(
       ForkChoiceStrategy forkChoiceStrategy, Bytes32 root, UnsignedLong slot) {
-    return forkChoiceStrategy
-        .blockSlot(root)
-        .flatMap(
-            blockSlot -> {
-              if (blockSlot.compareTo(slot) > 0) {
-                return get_ancestor(
-                    forkChoiceStrategy,
-                    forkChoiceStrategy.blockParentRoot(root).orElseThrow(),
-                    slot);
-              } else if (blockSlot.equals(slot)) {
-                return Optional.of(root);
-              } else {
-                // root is older than the queried slot, thus a skip slot. Return earliest root prior
-                // to slot.
-                return Optional.of(root);
-              }
-            });
+    Bytes32 parentRoot = root;
+    Optional<UnsignedLong> blockSlot = forkChoiceStrategy.blockSlot(root);
+    while (blockSlot.isPresent() && blockSlot.get().compareTo(slot) > 0) {
+      parentRoot = forkChoiceStrategy.blockParentRoot(parentRoot).orElseThrow();
```

Comment:
This looks good to me, but independent from this PR, makes me wonder if throwing here when the parentRoot does not exist in our ProtoArray was the right call. Seems like we simply should be returning an empty optional when we don't have the parent root, since the parent root is potentially a root of a block in a non-finalized fork, and we should be fine with that. Will do a follow-up PR on this.

Reply:
Actually throwing there does make sense because, for a block that we found the slot in the ProtoArray, we should have the complete ProtoNode anyway.